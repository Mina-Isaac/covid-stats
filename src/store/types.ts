export interface CaseDataPoint {
  cases: number;
  date: string;
}

export interface IncidenceDataPoint {
  weekIncidence: number;
  date: string;
}

interface APIResponse<
  AGS extends string,
  DataPoint extends CaseDataPoint | IncidenceDataPoint
> {
  data: {
    [key in AGS]: {
      ags: AGS;
      name: string;
      history: DataPoint[];
    };
  };
  meta: {
    source: string;
    contact: string;
    info: string;
    lastUpdate: string;
    lastCheckedForUpdate: string;
  };
  error: never;
}

export interface ErrorResponse {
  code: number;
  error: string;
  message: string;
  data: never;
}

export type CaseAPIResponse = APIResponse<string, CaseDataPoint>;
export type IncidenceAPIResponse = APIResponse<string, IncidenceDataPoint>;
