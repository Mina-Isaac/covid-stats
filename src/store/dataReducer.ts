import {
  createAsyncThunk,
  createReducer,
  SerializedError,
} from "@reduxjs/toolkit";
import { RootState } from ".";
import { CaseAPIResponse, IncidenceAPIResponse, ErrorResponse } from "./types";
import service from "../services";

export interface DataState {
  caseData: CaseAPIResponse | ErrorResponse | undefined;
  caseDataStatus: "idle" | "loading" | "failed" | "succeeded";

  incidenceData: IncidenceAPIResponse | ErrorResponse | undefined;
  incidenceDataStatus: "idle" | "loading" | "failed" | "succeeded";

  error: SerializedError | string | undefined;
}

const initialState: DataState = {
  caseData: undefined,
  caseDataStatus: "idle",

  incidenceData: undefined,
  incidenceDataStatus: "idle",

  error: undefined,
};

export const fetchCaseData = createAsyncThunk(
  "data/cases",
  async (_: void, { getState, rejectWithValue }) => {
    const {
      settings: { AGS, dayCount },
    } = getState() as RootState;

    if (AGS) {
      const response = await service.fetchCaseData({ AGS, dayCount });
      if (response?.error) return rejectWithValue(response.error);
      return response;
    }
  }
);

export const fetchIncidenceData = createAsyncThunk(
  "data/incidence",
  async (_: void, { getState, rejectWithValue }) => {
    const {
      settings: { AGS, dayCount },
    } = getState() as RootState;

    if (AGS) {
      const response = await service.fetchIncidenceData({ AGS, dayCount });
      if (response?.error) return rejectWithValue(response.error);
      return response;
    }
  }
);

const dataReducer = createReducer(initialState, (builder) => {
  builder
    .addCase(fetchCaseData.pending, (state) => {
      state.caseDataStatus = "loading";
      state.error = undefined;
    })

    .addCase(fetchCaseData.fulfilled, (state, { payload }) => {
      state.caseDataStatus = "succeeded";
      state.caseData = payload;
    })

    .addCase(fetchCaseData.rejected, (state, action) => {
      state.caseDataStatus = "failed";
      state.error = action.payload as string || action.error;
    })
    .addCase(fetchIncidenceData.pending, (state) => {
      state.incidenceDataStatus = "loading";
    })
    .addCase(fetchIncidenceData.fulfilled, (state, { payload }) => {
      state.incidenceDataStatus = "succeeded";
      state.incidenceData = payload;
    })
    .addCase(fetchIncidenceData.rejected, (state, action) => {
      state.incidenceDataStatus = "failed";
      state.error = action.payload as string || action.error;
    });
});

// Selectors related to data state

const selectCaseHistory = ({
  data,
  settings: { AGS: districtAGS },
}: RootState) => districtAGS ? data?.caseData?.data?.[districtAGS]?.history: undefined;

const selectIncidenceHistory = ({
  data,
  settings: { AGS: districtAGS },
}: RootState) =>
  districtAGS ? data.incidenceData?.data?.[districtAGS]?.history : undefined;

const selectCaseDataStatus = (state: RootState) => state.data.caseDataStatus;
const selectIncidenceDataStatus = (state: RootState) =>
  state.data.incidenceDataStatus;

const selectError = (state: RootState) => state.data.error;

export {
  selectCaseHistory,
  selectIncidenceHistory,
  selectCaseDataStatus,
  selectIncidenceDataStatus,
  selectError,
};

export default dataReducer;
