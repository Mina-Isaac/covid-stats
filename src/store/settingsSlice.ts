import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from ".";

interface SettingsState {
  AGS: string | undefined;
  dayCount: number;
  drawnStats: {
    cases: boolean;
    incidence: boolean;
  };
}

const initialState: SettingsState = {
  AGS: "09162",
  dayCount: 90,
  drawnStats: {
    cases: true,
    incidence: true,
  },
};

const settingsSlice = createSlice({
  name: "settings",
  initialState,
  reducers: {
    setDistrictAGS: (state, { payload }: PayloadAction<string>) => {
      state.AGS = payload;
    },

    setDayCount: (state, { payload }: PayloadAction<number>) => {
      state.dayCount = payload;
    },

    setDrawnStats: (
      state,
      { payload }: PayloadAction<keyof SettingsState["drawnStats"]>
    ) => {
      state.drawnStats[payload] = !state.drawnStats[payload];
    },
  },
});

export const { setDistrictAGS, setDayCount, setDrawnStats } =
  settingsSlice.actions;

// Settings state selectors
const selectDistrictAGS = (state: RootState) => state.settings.AGS;

const selectDayCount = (state: RootState) => state.settings.dayCount;

const selectDrawnStats = (state: RootState) => state.settings.drawnStats;

export { selectDistrictAGS, selectDayCount, selectDrawnStats };

export default settingsSlice.reducer;
