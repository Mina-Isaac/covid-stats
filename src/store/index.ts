import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import dataReducer from "./dataReducer";
import settingsReducer from "./settingsSlice";

export const store = configureStore({
  reducer: {
    data: dataReducer,
    settings: settingsReducer,
  },
  devTools: process.env.NODE_ENV === "development",
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
