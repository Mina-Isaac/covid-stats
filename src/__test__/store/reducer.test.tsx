import { RootState} from "../../store/";
import dataReducer, {
  fetchCaseData,
} from "../../store/dataReducer";
import settingsReducer, {
  setDayCount,
  setDistrictAGS,

} from "../../store/settingsSlice";

function getRandomInt(min: number = 90, max: number = 2500) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

const getInitialDataState = (initial?: RootState["data"]) =>
  dataReducer(initial, {} as any);

const getInitialSettingsState = (initial?: RootState["settings"]) =>
  settingsReducer(initial, {} as any);

describe("Reducer data flow", () => {
  it("should match a snapshot", () => {
    const initialState = getInitialDataState();
    expect(initialState).toMatchSnapshot();
  });

  it("store data on contacts slice on fetchCaseData.fulfilled", () => {
    const initialState = getInitialDataState();
    const sampleResponse = { someKey: "someValue" };
    const state = dataReducer(
      initialState,
      //@ts-expect-error
      fetchCaseData.fulfilled(sampleResponse)
    );
    expect(state.caseData).toBe(sampleResponse);
    expect(state.error).toEqual(undefined);
    expect(state.caseDataStatus).toEqual("succeeded");
  });

  it("sets district AGS", () => {
    const initialState = getInitialSettingsState();
    const someAGS = "012345";
    const state = settingsReducer(initialState, setDistrictAGS(someAGS));
    expect(state.AGS).toEqual(someAGS);
  });

  it("sets selected contact", () => {
    const initialState = getInitialSettingsState();
    const someRandomNumber = getRandomInt()
    const state = settingsReducer(
      initialState,
      setDayCount(someRandomNumber)
    );
    expect(state.dayCount).toBe(someRandomNumber);
  });
});
