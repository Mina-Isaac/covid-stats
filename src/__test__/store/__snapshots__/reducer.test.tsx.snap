// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer data flow should match a snapshot 1`] = `
Object {
  "caseData": undefined,
  "caseDataStatus": "idle",
  "error": undefined,
  "incidenceData": undefined,
  "incidenceDataStatus": "idle",
}
`;
